### PR TITLE
Commerce 1067 3

### PR DESCRIPTION
**Pull request by shuyangzhou**

### CODE REVIEW

**Comment by brianchandotcom** [comment]
Merged. Thank you.
View total diff: https://github.com/brianchandotcom/liferay-portal/compare/f927e6b75b...6f0a28ee27

**Comment by brianchandotcom** [comment]
@ivicac see upstream for my changes.

File: modules/apps/batch-engine/batch-engine-service/src/main/java/com/liferay/batch/engine/internal/BatchEngineTaskMethodServiceTracker.java [hunk 1]
```diff
@@ -67,48 +66,90 @@ public void destroy() {
...
 
 	private final Map
-		<String,
+		<FactoryKey,
 		 UnsafeBiFunction
 			 <Company, User, BatchEngineTaskItemWriter,
 			  ReflectiveOperationException>>
 				_batchEngineTaskItemWriterFactories = new ConcurrentHashMap<>();
-	private final Map<Class<?>, List<String>> _itemClassKeysMap =
+	private final Map<String, Map.Entry<Class<?>, AtomicInteger>> _itemClasses =
```
Comment (ivicac) [inline modules/apps/batch-engine/batch-engine-service/src/main/java/com/liferay/batch/engine/internal/BatchEngineTaskMethodServiceTracker.java]: hi @shuyangzhou , quick question, when you said "add thread safety", does it mean that OSGI container is capable of in parallel registering/unregistering modules and components?

Comment (shuyangzhou) [reply modules/apps/batch-engine/batch-engine-service/src/main/java/com/liferay/batch/engine/internal/BatchEngineTaskMethodServiceTracker.java]: @ivicac the map is updated by a service tracker. When services are registered/unregistered, the caller thread is the one driving the call to the ServiceTracker. Anyone with access (directly or indirectly) to BundleContext has the ability to register/unregister service, so the ServiceTracker callback is very much possible to be executed concurrently.

Comment (ivicac) [reply modules/apps/batch-engine/batch-engine-service/src/main/java/com/liferay/batch/engine/internal/BatchEngineTaskMethodServiceTracker.java]: @shuyangzhou thx

**Pull request closed without merging**